Update jupyter lab (#5982)
galaxyproject/training-material: topics/galaxy-interface/tutorials/galaxy-intro-jupyter/tutorial.md

# What is Jupyter ?

> [Jupyter](http://jupyter.org/) is an interactive environment that mixes explanatory text, command line and output display for an interactive analysis environment. Its implementation in Galaxy facilitates the performance of additional analysis if there is no tool for it.

Reviewer: ```suggestion
[Jupyter](http://jupyter.org/) is an interactive environment that mixes explanatory text, command line and output display for an interactive analysis environment. Its implementation in Galaxy facilitates the performance of additional analyses if there are no tools for it.
```

# Use Jupyter notebook in Galaxy

## Import data

> To manipulate data first we upload the  [python-novice-gapminder-data.zip](https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip) folder into your Galaxy history. 

Reviewer: ```suggestion
To manipulate data, we first upload the [python-novice-gapminder-data.zip](https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip) folder into your Galaxy history. To add the files you can either upload them locally from your computer or use Zenodo.
```

To add the files you can either uploade them locally from your computer or use Zenodo.

> <hands-on-title>Data upload using Zenodo</hands-on-title>
>
> 1. Create a new history for this RNA-Seq exercise
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import the following tabular file from [Zenodo]({{ page.zenodo_link }}):
>    - `gapminder_all.csv` 
>    - `gapminder_gdp_africa.csv` 
>    - `gapminder_gdp_americas.csv`
>    - `gapminder_gdp_asia.csv`
>    - `gapminder_gdp_europe.csv`

> >    - `gapminder_gdp_oceania.csv`

Author: Your right! thanks!

>    {{ page.zenodo_link }}/files/gapminder_gdp_americas.csv
>    {{ page.zenodo_link }}/files/gapminder_gdp_asia.csv
>    {{ page.zenodo_link }}/files/gapminder_gdp_europe.csv
>    {{ page.zenodo_link }}/files/gapminder_gdp_oceania.csv
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
>
> 3. Make sure the file is imported as `CSV` by expanding the box of each imported file in your history and check the format.
> 
> If the format is not correct change the format. 
>    - Click on the pencil icon {% icon galaxy-pencil %}
>    - Select the `Datatype`  {% icon galaxy-chart-select-data %} tab in the middle panel
>    - In `Assign Datatype` section: change the data format to `csv` and click on `save`
>
{: .hands_on}

## Open the JupyterLab environment

Opening up your {% tool [JupyterLab](interactive_tool_jupyter_notebook) %}:

{% snippet faqs/galaxy/interactive_tools_jupyter_launch.md %}

You should now be looking at a page with the JupyterLab interface:

![Jupyterlab default session](../../images/jupyterlab_default_session.png)

As shown on the figure above, JupyterLab interface is made of 3 main areas:
- The menu bar at the top

> - The left side bar with in particular the **File Browser**

Reviewer: ```suggestion
- The left side bar with, in particular, the **File Browser**
```

- The main work area in the central panel

> The Jupyter notebook can be started from different points. You can either open a Jupyter notebook from a dataset in your history or from the *Visualize* tab in the upper menu.

Reviewer: I think this would be better in the "Hands on" section

## Start your first notebook

> Now that we are ready to start exploring JupyterLab, and open a python Notebook. There will be one pre-opened notebook available in the file browser on the left side called **ipython_galaxy_notebook.ipynb**. But we will open a new Jupyter notebook, select a kernel and give it a name.

Reviewer: ```suggestion
Now that we are ready to start exploring JupyterLab, let's open a python Notebook. There will be one pre-opened notebook available in the file browser on the left side called **ipython_galaxy_notebook.ipynb**. For this training, however, we will open a new Jupyter notebook, select a kernel and give it a name.
```

> <hands-on-title>Start a notebook</hands-on-title>
> 
> 1. Open a new Jupyter notebook 
> 
>    There is more than one option to open a jupyter notebook. One option is:   
>  
>    - Click on **File** in the top menu bar
>    - Select **New** -> **Notebook**
>    - Choose the kernel `Python [conda env:python-kernel-3.12]` from the dropdown menu and click **select**
>
> 2. Change the kernel
>

> >    If the kernel `Python [conda env:python-kernel-3.12]` was chosen in the previous step it should appear in the upper right corner of the notebook file. If not this is the location where the kernel can be switched. 

Reviewer: ```suggestion
>    If the kernel `Python [conda env:python-kernel-3.12]` was chosen in the previous step, it should appear in the upper right corner of the notebook file. If not, this is the location where the kernel can be switched. 
```

>  
>    - Click on the field that displays the current kernel e.g. `Python [conda env:base]`
>    - Now select the kernel `Python [conda env:python-kernel-3.12]` from the drop down menu
>
> 3. Name the Jupyter notebook
>
>    There are several options how to name or rename your Jupyter Notebook. One way is to: 
>  
>    - Click on **File** and select **Save File As...**
>    - Enter your file name e.g. `first_galaxy_notebook.ipynb`. Note that your file needs to end with `.ipynb`
>    - Click on **Save**
>
{: .hands_on}

## Install and import libraries using Conda

Some dependencies or programming libraries may not be available in the kernel your Jupyter Notebook is using. If that's the case, you can install or update these libraries using **Conda**.

> <hands-on-title>Install from a Conda recipe</hands-on-title>
>
> 1. Click on a cell of your notebook to edit it (verify that it is defined as a `Code` cell)
> 2. Enter the following lines : 
>    ```text
>!conda install -y pandas
>!conda install -y seaborn
>    ```

> >   - The `!` indicate you are typing a bash command line (alternatively you can use `%%bash` at the beginning of your cell )

Reviewer: ```suggestion
>   - The `!` indicate you are typing a bash command line (alternatively you can add the line `%%bash` at the beginning of your cell. In that case the whole cell will be run as bash commands.)
```

>   - The `-y` option allows the installation without asking for confirmation  (The confirmation is not managed well by notebooks)
> 3. `shift`+`return` to run the cell or click on the run cell button.
>
{: .hands_on}

Now you will be able to import this Python libraries and use them with your library code. 

> <hands-on-title>Install Python libraries using Conda</hands-on-title>
>
> 1. Click on a cell of your notebook to edit it (verify that it is defined as a `Code` cell)
>
> 2. Enter the following lines : 
>    ```python
>import pandas as pd
>import seaborn as sns
>from IPython.display import display
>import matplotlib.pyplot as plt
>    ```
> 3. `shift`+`return` to run the cell or click on the run cell button.
>
{: .hands_on}

> If you wish to follow the Software Carpentries training [Plotting and Programming in Python](https://swcarpentry.github.io/python-novice-gapminder/instructor/index.html) after finishing this Jupyter Notebook introduction you should install the following Python libraries.

Reviewer: ```suggestion
If you wish to follow the Software Carpentries training [Plotting and Programming in Python](https://swcarpentry.github.io/python-novice-gapminder/instructor/index.html) after finishing this Jupyter Notebook introduction, you should install the following Python libraries.
```

> <hands-on-title>Install Python libraries for a Python introduction</hands-on-title>
>

> > 1.  Copy the following install commands and pasted them into a empty cell of your Notebook

Reviewer: ```suggestion
> 1.  Copy the following install commands and paste them into a empty cell of your Notebook
```

>    ```text
>!conda install -y math
>!conda install -y matplotlib
>!conda install -y glob
>!conda install -y pathlib
>    ```
>
> 2. press `shift`+`return` to run the cell or click on the run cell button.
>
{: .hands_on}

## Import data 
If you want to include datasets from your history into your  Jupyter notebook, you can import them  using the `get(12)` command, with the number of your dataset in the history (If you are working on a collection, unhide datasets to see their numbers).
You can use the `gapminder_gdp_europe.csv` file. You can save all of the files into a path location first, to later refer to this file using the import path variable.

> > <hands-on-title>Import file location from your history</hands-on-title>

Reviewer: ```suggestion
> <hands-on-title>Import a file location from your history</hands-on-title>
```

>
> Save the import file location to a variable (`file_import`) name with the `get()` function.
>    ```python
>file_import = get("[file_number]")
>    ```
>    - The files are referenced in Jupyter by their number in the history.
>    - The variable `file_import` now stores the location where your file can be imported from.
>
{: .hands_on}

If you wish to follow the python training later you should import all of the gapminder datasets now. The `file_import` variables can be treated like the path with the tutorial. 

> > <hands-on-title>Import file location from your history</hands-on-title>

Reviewer: ```suggestion
> <hands-on-title>Import the training data locations from your history</hands-on-title>
```

>
> 1. Find the `file_numbers` of all of the gapminder datasets (`gapminder_all.csv`, `gapminder_gdp_africa.csv`, `gapminder_gdp_americas.csv`, `gapminder_gdp_asia.csv`, `gapminder_gdp_europe.csv`, `gapminder_gdp_oceania.csv`) with your galaxy history
> 2. Load the datasets `file_import`into your Ipython notebook:
>    ```python
>gapminder_all_import = get("[file_number]")
>gapminder_gdp_africa_import = get("[file_number]")
>gapminder_gdp_americas_import = get("[file_number]")
>gapminder_gdp_asia_import = get("[file_number]")
>gapminder_gdp_europe_import = get("[file_number]")
>gapminder_gdp_oceania_import = get("[file_number]")
>    ```
>    - The files are referenced in Jupyter by their number in the history.
>    - The variable `file_import` now stores the location where you file can be imported from
>
{: .hands_on}

## Graph Display in Jupyter

> In this tutorial we are going to plot a distribution graph of our data. For this we will first need to load one of our tabular data files. You can use the `gapminder_all.csv` file.

Reviewer: ```suggestion
In this tutorial we are going to plot a distribution graph of our data. For this, we will first need to load one of our tabular data files. You can use the `gapminder_all.csv` file.
```

## Export Data

If you want to save a file you generated in your notebook, use the `put("file_name")` command. That is what we are going to do with our distribution plot.

> <hands-on-title>Save an Jupyter-generated image into a Galaxy History</hands-on-title>
>
> 1. Create an image file with the figure you just drew with the command
>    ```python
>fig.savefig('distplot.png')
>    ```
> 2. Export your image into your history with the command  
>    ```python
>put('distplot.png')
>    ``` 
>
{: .hands_on}

## Save the Notebook in your history

Once you are done with your analysis or anytime during the editing process, you can save the notebook into your history using the `put("first_galaxy_notebook.ipynb")`. If you create additional notebooks with different names, make sure you save them all before you quit JupyterLab.

This will create a new notebook `.ipynb` file in your history every time you click on this icon.

> <hands-on-title>Closing JupyterLab</hands-on-title>
>
> 1. In the Galaxy interface click on  **Interactive Tools** button on the left side.
>
> 2. Tick {% icon galaxy-selector %} the box of your Jupyter Interactive Tool, and click **Stop**.
>
{: .hands_on}

> If you want to reopen your Jupyter Notebook again you will start a new **Jupyter Interactive Tool** and 

Reviewer: ```suggestion
If you want to reopen a Jupyter Notebook saved in your history, you can use the tool  **Interactive JupyterLab Notebook**, select "Load a previous Notebook", and select the notebook from your history. 
```

# Use Jupyter notebook in Galaxy

## Import data
To manipulate data first we upload the  [python-novice-gapminder-data.zip](https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip) folder into your Galaxy history. 

> <hands-on-title>Data upload</hands-on-title>
>
> 1. Create a new history for this tutorial
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import [python-novice-gapminder-data.zip](https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip) folder. The folder contains the following tabular files:
>    - `gapminder_all.csv` 
>    - `gapminder_gdp_africa.csv` 
>    - `gapminder_gdp_americas.csv`
>    - `gapminder_gdp_asia.csv`
>    - `gapminder_gdp_europe.csv`
>    - `gapminder_gdp_oceania.csv`
>
>    ```text
>    https://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
>
>

> > 3. Unzip data using the {% tool [Unzip](toolshed.g2.bx.psu.edu/repos/imgteam/unzip/unzip/6.0+galaxy0) %}

Reviewer 2: We could safe this step if we just re-upload the data to Zenodo

Author: Thanks for the suggestion. I did add data to Zenodo and updated the tutorial. If @bgruening, @Delphine-L or @shiltemann have some extra time I would appreciate a review. Thanks a lot!
